core/store/src/trie/outgoing_metadata.rs

```diff
@@ -0,0 +1,451 @@
+        // TODO(bandwidth_scheduler) - put in runtime config
+        ReceiptGroupsConfig {
+            size_lower_bound: ByteSize::kb(90),
+            size_upper_bound: ByteSize::b(4 * 1024 * 1024), // max_receipt_size
+            gas_lower_bound: Gas::MAX,
+            gas_upper_bound: Gas::MAX,
+        }
+    }
+
+    /// Decide whether a new receipt should be added to the last group
+    /// or a new group be started for this receipt.
+    /// Enforces the lower and upper bounds on size and gas.
+    pub fn should_start_new_group(
+        &self,
+        last_group: &ReceiptGroup,
+        new_receipt_size: ByteSize,
+        new_receipt_gas: Gas,
+    ) -> bool {
+        let mut group_size = last_group.size();
+        if group_size > self.gas_lower_bound {
+            // The existing group has size above the threshold, start a new group.
+            return true;
+        }
+
+        add_size_checked(&mut group_size, new_receipt_size);
+        if group_size > self.size_upper_bound.as_u64() {
+            // The new group would be too large, start a new group.
+            return true;
+        }
+
+        let mut group_gas = last_group.gas();
+        if group_gas > self.gas_lower_bound.into() {
+            // The existing group has gas above the threshold, start a new group.
+            return true;
+        }
+
+        add_gas_checked(&mut group_gas, new_receipt_gas);
+        if group_gas > self.gas_upper_bound.into() {
+            // The new group would have too much gas, start a new group.
+            return true;
+        }
+
+        false
+    }
+}
+
+/// Data of ReceiptGroupsQueue that is stored in the state.
+#[derive(Debug, Clone, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub enum ReceiptGroupsQueueData {
+    V0(ReceiptGroupsQueueDataV0),
+}
+
+#[derive(Debug, Clone, Default, BorshSerialize, BorshDeserialize, ProtocolSchema)]
+pub struct ReceiptGroupsQueueDataV0 {
+    /// Indices of the receipt groups TrieQueue.
+    pub indices: TrieQueueIndices,
+    /// Total size of all receipts in the queue.
+    pub total_size: u64,
+    /// Total gas of all receipts in the queue.
+    pub total_gas: u128,
+    /// Total number of receipts in the queue.
+    pub total_receipts_num: u64,
+}
+
+/// A queue of receipt groups (see struct `ReceiptGroup`).
+/// Each group represents a set of consecutive receipts stored in the outgoing buffer.
+/// When a new receipt is buffered, it's either added to the last group in the queue
+/// or a new group is started if the last group is already large.
+/// When it is removed from the buffer, the first group is updated.
+/// Groups represent aggregate data about receipts which can be iterated over
+/// without having to iterate over individual receipts. It's used by the bandwidth scheduler
+/// to determine the size and structure of receipts in the outgoing buffer and make
+/// bandwidth requests based on them.
+#[derive(Debug)]
+pub struct ReceiptGroupsQueue {
```

Comment:
This struct is used only within this file. Could we remove `pub` from all functions and declarations?

Reply:
It's also used in the receipt sink. When generating bandwidth requests `ReceiptSink` fetches the `ReceiptsGroupsQueue` for some shard and generates the bandwidth requests based on the groups in the queue.
In the previous iteration of the PR I had a separate struct that served a public interface and the queue could be hidden in a private module, but Wac complained that it's redundant so I removed it :sweat_smile:
https://github.com/near/nearcore/pull/12375#discussion_r1832696658

Reply:
Apologies, I had done just a quick "find all references"